wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs

```diff
@@ -35,107 +36,167 @@ data FeeOptions dom = FeeOptions {
     , foExpenseRegulation :: ExpenseRegulation
     }
 
+data CoinSelFinalResult dom = CoinSelFinalResult
+    { csrInputs  :: NonEmpty (UtxoEntry dom)
+    -- ^ Picked inputs
+    , csrOutputs :: NonEmpty (Output dom)
+    -- ^ Picked outputs
+    , csrChange  :: [Value dom]
+    -- ^ Resulting changes
+    }
+
+type PickUtxo m utxo
+    =  Value (Dom utxo)
+    -> CoinSelT utxo CoinSelHardErr m (Maybe (UtxoEntry (Dom utxo)))
+
 -- | Given the coin selection result from a policy run, adjust the outputs
 -- for fees, potentially returning additional inputs that we need to cover
 -- all fees.
-adjustForFees :: forall utxo m. (CoinSelDom (Dom utxo), Monad m)
-              => FeeOptions (Dom utxo)
-              -> (Value (Dom utxo) ->
-                   CoinSelT utxo CoinSelHardErr m (Maybe (UtxoEntry (Dom utxo))))
-              -> [CoinSelResult (Dom utxo)]
-              -> CoinSelT utxo CoinSelHardErr m
-                   ([CoinSelResult (Dom utxo)], SelectedUtxo (Dom utxo), Value (Dom utxo))
+adjustForFees
+    :: forall utxo m. (CoinSelDom (Dom utxo), Monad m)
+    => FeeOptions (Dom utxo)
+    -> PickUtxo m utxo
+    -> [CoinSelResult (Dom utxo)]
+    -> CoinSelT utxo CoinSelHardErr m (CoinSelFinalResult (Dom utxo))
 adjustForFees feeOptions pickUtxo css = do
-    case foExpenseRegulation feeOptions of
-      ReceiverPaysFee -> coinSelLiftExcept $
-        (, emptySelection, valueZero) <$> receiverPaysFee upperBound css
-      SenderPaysFee ->
-        senderPaysFee pickUtxo upperBound css
-  where
-    upperBound = feeUpperBound feeOptions css
+    let inps = concatMap (selectedEntries . coinSelInputs) css
```

Comment:
Suggested comment: We lose the relationship between the transaction outputs and their corresponding inputs/change outputs here. This is a decision we may wish to revisit later. For now however note that since (a) coin selection tries to establish a particular ratio between payment outputs and change outputs (currently it aims for an average of 1:1), and moreover (b) coin selection currently only generates a single change output per payment output, distributing the fee proportionally across all change outputs is roughly equivalent to distributing it proportionally over the payment outputs (roughly, not exactly, because the 1:1 proportion is best effort only, and may in some cases be wildly different). Note that for (a) we don't need the ratio to be 1:1, the above reasoning will remain true for any proportion `1:n`. For (b) however, if coin selection starts creating _multiple_ outputs, _and_ this number may _vary_, then losing the connection between outputs and change outputs will mean that that some outputs may pay a larger percentage of the fee (depending on how many change outputs the algorithm happened to choose).